### PR TITLE
fix(internals): Change isCi to check process.stdin TTY status

### LINKED ISSUE #14620: Prisma CLI non-interactive detection is incorrect.
### Bug description

The `prisma migrate` CLI command guards against use in non-interactive environments, but it can lead to false positives.

### How to reproduce

Run `prisma migrate` from a CLI helper tool like `oclif` using `zx` to execute the command. It is capable of receiving input, but Prisma thinks it's non-interactive.

### Expected behavior

`prisma migrate` doesn't error with the non-interactive error.

### Prisma information

Not relevant, applies to any schema

### Environment & setup

n/a

### Prisma Version

```
prisma                  : 4.1.1
@prisma/client          : 4.1.1
Current platform        : darwin
Query Engine (Node-API) : libquery-engine 8d8414deb360336e4698a65aa45a1fbaf1ce13d8 (at node_modules/@prisma/engines/libquery_engine-darwin.dylib.node)
Migration Engine        : migration-engine-cli 8d8414deb360336e4698a65aa45a1fbaf1ce13d8 (at node_modules/@prisma/engines/migration-engine-darwin)
Introspection Engine    : introspection-core 8d8414deb360336e4698a65aa45a1fbaf1ce13d8 (at node_modules/@prisma/engines/introspection-engine-darwin)
Format Binary           : prisma-fmt 8d8414deb360336e4698a65aa45a1fbaf1ce13d8 (at node_modules/@prisma/engines/prisma-fmt-darwin)
Default Engines Hash    : 8d8414deb360336e4698a65aa45a1fbaf1ce13d8
Studio                  : 0.469.0
```

### PR DESCRIPTION
**Pull request by AndrewSouthpaw**
Fixes #14620.

## Test plan

I was unable to run the tests locally, it was giving me an error:

```
$ pnpm run test

> dev-env@ test /Users/andrew/Projects/prisma
> ts-node scripts/ci/publish.ts --test

Error: Error: Missing env var BUILDKITE_BRANCH
    at publish (/Users/andrew/Projects/prisma/scripts/ci/publish.ts:509:11)
```

So I'm hoping the CI will run the tests and we can see what breaks...

### CODE REVIEW

**Comment by AndrewSouthpaw** [comment]
Hmm I'm totally unfamiliar with Buildkite, and am rather surprised by the breadth of errors that seem to be showing up in tests. Could someone speak to whether this is really all because of the change to `isCi`? Or is there something else I did wrong with the PR?

**Comment by Jolg42** [comment]
@AndrewSouthpaw I think these errors are unrelated, I'll merge main, and we can see what happens.

**Comment by AndrewSouthpaw** [comment]
Thanks @Jolg42! It looks like this [test is failing](https://github.com/prisma/prisma/runs/7670747820?check_suite_focus=true#step:13:761). I followed the instructions in `TESTING.md` but couldn't get the tests to work correctly. Do you have any advice on running the tests locally so I can start looking into fixing the test?

**Comment by Joe-Kollin** [comment]
Hey if you need a quick fix a library like `patch-package` will allow you to monkey patch the isCi() method. I had to do the same thing because I was getting an error. Not a great long term fix, but it's nice if you're in a bind

**Comment by Jolg42** [comment]
@AndrewSouthpaw From what I see this is one of our flaky test, so nothing to fix.

Though, since all tests are passing, we would need to check what this changes and add a test for it.

**Comment by AndrewSouthpaw** [comment]
Awesome! Happy to make that happen. Unfortunately, I'm having trouble running the tests locally, I get an error like this:

```
$ pnpm run test

> dev-env@ test /Users/andrew/Projects/prisma
> ts-node scripts/ci/publish.ts --test

Error: Error: Missing env var BUILDKITE_BRANCH
    at publish (/Users/andrew/Projects/prisma/scripts/ci/publish.ts:509:11)
```

The TESTING.md doc is thorough but didn't help me figure this issue out. If you wouldn't mind pointing me in the right direction, I'd be happy to puzzle over designing a good test for it.

**Comment by AndrewSouthpaw** [comment]
Turns out I was misreading the instructions, one of them mentioned `pnpm run jest` which led me down the wrong path. I found I could run jest easily from my IDE (WebStorm) or by running `pnpm test ...`. I couldn't get all tests to run locally though. Hopefully CI works fine 🤞🏼 

I added unit tests for `isCi`, LMK if you'd like the utils to go somewhere, I didn't see anything like a "test utils" file or folder on cursory inspection.

The integration test with CLI is covered [here](https://github.com/prisma/prisma/blob/8e90f8a1ca12f9b596f92bae811d9d92266aab7a/packages/migrate/src/__tests__/MigrateDev.test.ts#L134-L145), which upon closer inspection has both `process.stdout.isTTY` and `process.stdin.isTTY` as undefined so it shakes out. I didn't quite understand what was going on with the fixture and how the tests end up mocking an interactive environment, but it seems like maybe I didn't need to figure that part out.

Let me know what you think!

**Comment by AndrewSouthpaw** [comment]
Ok looks like there was some tweaks to be done to get the tests passing in CI, we should be good to go for review.

**Comment by AndrewSouthpaw** [comment]
Hi there! Just wanted to bump this one, any feedback I could address or things I could work on to help move this PR along?

File: packages/internals/src/utils/__tests__/isCi.test.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  })
+  afterEach(() => {
+    // If you set undefined in process.env it stringifies it, so we
+    // need special handling for that case of "unsetting" the process.env.
+    if (object === process.env && original === undefined) {
+      delete object[prop]
+    } else {
+      object[prop] = original
+    }
+  })
```
[the remark this answers is not in the document]
Comment (AndrewSouthpaw) [reply packages/internals/src/utils/__tests__/isCi.test.ts:32]: Ah sure good call

**Review by AndrewSouthpaw — COMMENTED** [review]
Thanks for the feedback! Updated.

**Comment by AndrewSouthpaw** [comment]
Hmm it seems like [this check](https://github.com/prisma/prisma/runs/7911812531?check_suite_focus=true) is failing for unrelated reasons, and then [this check](https://github.com/prisma/prisma/runs/7911812458?check_suite_focus=true) failed but ran fine locally. I've bumped CI to see if it happens again. Strange because everything was running fine before I added the last commit refactoring the code but not changing behavior at all.

**Comment by Jolg42** [comment]
@AndrewSouthpaw we do have some flaky tests... and once in a while someone catches a few of them at the same time. It looks like the PR passes all the tests at the moment!

**Comment by AndrewSouthpaw** [comment]
Woohoo! First commit to Prisma! 🎉 

Thanks for the feedback and helping it across the finish line. I'll be excited to see it in 4.3.0!

**Review by Jolg42 — APPROVED** [review]
Thank you for the issue and the PR 💚 

I'll merge now and do some extra testing.

Note: I think we should move out "is it interactive?" checks from `isCi`. It's pretty weird right now.

**Comment by AndrewSouthpaw** [comment]
>  I think we should move out "is it interactive?" checks

Yeah, it did surprise me a bit. I'd be happy to do some refactoring if you think it was warranted.

From what you wrote in the issue:

> Checking the codebase, why do we use isCi?
> - handlePanic: to skip the "send us the crash logs" prompt
> - db push: to throw with CLI flags to use
> - migrate dev: to throw environment is non-interactive, which is not supported
> - migrate reset: to throw environment is non-interactive, which is not supported

It seems like there's two separate questions, one is `isCI` and the other is `isInteractive`. If it doesn't make sense to justify behavior in the first two options with asking the question "is it interactive", then yeah I'd say splitting apart the function makes sense.

It _seems_ that at least the first one, handlePanic, makes sense to not do when non-interactive, since how would they respond?

For `db push`, I didn't quite understand what you meant by "throw with CLI flags to use".

So... depending on what you're looking for, I could even see an argument for dropping `isCI` entirely and just focusing on `isInteractive`.

**Comment by Jolg42** [comment]
@AndrewSouthpaw so here is the PR https://github.com/prisma/prisma/pull/15008

TLDR:
We still want to throw errors if standard input is not a TTY (= non interactive) or if running in CI:
If not interactive: we cannot prompt, prompt will not work (since https://github.com/prisma/prisma/pull/14649 we now check `stdin` and not `stdout`)
if running in CI: the prompt will hang the process until a timeout kicks in

**Comment by Jolg42** [comment]
>For db push, I didn't quite understand what you meant by "throw with CLI flags to use".
Here https://github.com/prisma/prisma/blob/fea4a6dd826361d8aade6d6481e746cc393499f0/packages/migrate/src/commands/DbPush.ts#L217 is I think what I thought. 
We throw with a message saying "use a flag", normally we woult prompt the user, but if we can't we ask to pass an explicit flag to the command.
